MAISTRA-2231: Seed MemberRoll listeners with system namespace

bison opened the pull request:
The MemberRoll controller was changed in MAISTRA-2197 to not send the
initial update with just the system namespace to all listeners.  This
can be a problem when the user hasn't created an SMMR resource yet,
because any listeners registered before the caches sync may never get
an update informing them to watch the system namespace.

In Maistra 2.1 this actually prevents the full install from completing
in some cases, e.g. the namespace controller may never create the
ConfigMap containing the CA certificate, which prevents the gateway
pods from starting.

This attempts to work around the issue by having all listeners
individually wait on the caches to sync, then seed just the system
namespace if no SMMR resource is found.

dgn commented on pkg/servicemesh/controller/controller.go:
```diff
@@ -121,19 +107,38 @@ func (smmrc *serviceMeshMemberRollController) newServiceMeshMemberRollListener(l
-			smmr := item.(*v1.ServiceMeshMemberRoll)
-			members = smmr.Status.ConfiguredMembers
+			smmrLog.Infof("Seeding listener %q with system namespace.", name)
+			handler.updateNamespaces("seed", smmrc.memberRollName, smmrc.getNamespaces(nil))
```
you should be able to call this with `members: nil`; the `updateNamespaces` func calls `getNamespaces` anyway

bison replied:
Thanks. Fixed.

dgn commented on pkg/servicemesh/controller/controller.go:
```diff
@@ -121,19 +107,38 @@ func (smmrc *serviceMeshMemberRollController) newServiceMeshMemberRollListener(l
+	// initial update only if the expected SMMR is not found in the cache.
+	// Unfortunately, there is a race here between sending this initial update
+	// and a user creating an SMMR resource.
+	go func() {
```
this will now create one goroutine per listener that'll poll the `HasSynced()` func... not sure if that's ideal. It does avoid having to use a mutex, but I'm not sure about the trade-off there. before the cache is synced, we're now creating ~20 goroutines all hammering the SMMR informer. it's probably not a big deal, but do you think it's better than a single goroutine + mutex?

bison replied:
I see what you mean, but I don't expect it to be a problem in practice. It's not that many threads, and they are polling with a reasonable interval, and the `HasSynced()` method on informers is not doing any heavy lifting -- it's not hitting the API or anything. This seemed fine and easier to read to me since it's all in one place.

dgn replied:
yeah I agree, it's not a ton of goroutines and they really don't do much

jwendell replied:
I prefer the single goroutine+lock here, but since it's a temporary fix, that's ok.

bison replied:
This uses a flag and a RW mutex now, though it's not a single go routine. We still have to poll in the listeners.

dgn reviewed (approved):
lgtm, thanks! For a longer-term solution, I think your idea of having an Add/Remove interface instead of SetNamespaces would be worth exploring, if people agree I'll file a ticket and we can look into it for 2.1

jwendell commented on pkg/servicemesh/controller/controller.go:
```diff
@@ -121,19 +107,38 @@ func (smmrc *serviceMeshMemberRollController) newServiceMeshMemberRollListener(l
-	if smmrc.cacheWarmed {
-		smmrLog.Debugf("Listener for %q created. Ready to send an initial update", name)
+	// WARNING: This is a bit of a hack.  It is mostly intended as a temporary
+	// workaround to enable 2.1 testing to begin.
```
Please, file an issue in 2.1 bucket and link it here

bison replied:
Updated to match what is in 2.0 -- this was already for 2.1 -- don't think there's anything to do here.

dgn commented on pkg/servicemesh/controller/controller.go:
```diff
@@ -121,19 +107,38 @@ func (smmrc *serviceMeshMemberRollController) newServiceMeshMemberRollListener(l
-		for _, item := range smmrc.informer.GetIndexer().List() {
-			smmr := item.(*v1.ServiceMeshMemberRoll)
-			members = smmr.Status.ConfiguredMembers
+			smmrLog.Infof("Seeding listener %q with system namespace.", name)
```
nit: the `updateNamespaces` func is already logging (at `debug` level), do we need this log line?

bison replied:
Don't have a strong opinion, but this is only logged in the case where the SMMR doesn't exist. Seemed kind of nice to know without having to up the log level. Happy to make it debug or remove it though.

dgn commented:
/test unit

jwendell commented:
likely the timeout on that test needs to be increased... prow is too slow... it takes 2 seconds on my laptop. 1 minute on prow is not enough.

bison commented:
Going to pause this until we make a decision on #308.

bison committed:
MAISTRA-2234: Seed MemberRoll listeners with system namespace

The MemberRoll controller was changed in MAISTRA-2197 to not send the
initial update with just the system namespace to all listeners.  This
can be a problem when the user hasn't created an SMMR resource yet,
because any listeners registered before the caches sync may never get
an update informing them to watch the system namespace.

This attempts to work around the issue by having all listeners
individually wait on the caches to sync, then seed just the system
namespace if no SMMR resource is found.

bison commented:
We merged #308 into the 2.0 branch, so I've updated this to match it. I think this is ready now.
